src/tango/ip/fd_ip.c

```diff
@@ -0,0 +1,381 @@
+#include "fd_ip.h"
+
+#include <arpa/inet.h>
+
+ulong
+fd_ip_align( void ) {
+  return FD_IP_ALIGN;
+}
+
+
+ulong
+fd_ip_footprint( ulong arp_entries,
+                 ulong route_entries ) {
+
+  ulong l;
+
+  l = FD_LAYOUT_INIT;
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, sizeof(fd_ip_t)                             );
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, sizeof(fd_nl_t)                             );
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, arp_entries   * sizeof(fd_nl_arp_entry_t)   );
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, route_entries * sizeof(fd_nl_route_entry_t) );
+
+  return FD_LAYOUT_FINI( l, FD_IP_ALIGN );
+}
+
+
+void *
+fd_ip_new( void * shmem,
+           ulong  arp_entries,
+           ulong  route_entries ) {
+  if( !fd_ulong_is_aligned( (ulong)shmem, FD_IP_ALIGN ) ) {
+    FD_LOG_ERR(( "Attempt to fd_ip_new with unaligned memory" ));
+    return NULL;
+  }
+
+  ulong l;
+  uchar * mem = (uchar*)shmem;
+
+  l = FD_LAYOUT_INIT;
+
+  fd_ip_t * ip = (fd_ip_t*)mem;
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, sizeof(fd_ip_t)                             );
+
+  ulong ofs_netlink = l;
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, sizeof(fd_nl_t)                             );
+
+  ulong ofs_arp_table   = FD_ULONG_ALIGN_UP( l, FD_IP_ALIGN );
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, arp_entries   * sizeof(fd_nl_arp_entry_t)   );
+
+  ulong ofs_route_table = FD_ULONG_ALIGN_UP( l, FD_IP_ALIGN );
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, route_entries * sizeof(fd_nl_route_entry_t) );
+
+  ulong mem_sz = FD_LAYOUT_FINI( l, FD_IP_ALIGN );
+
+  /* clear all to zero */
+  fd_memset( ip, 0, mem_sz );
+
+  /* set values in ip */
+  ip->magic                = FD_IP_MAGIC;
```

Comment:
Can align this to the pattern used elsewhere where magic is set last, so if it matches you know the struct has been initialized. Also with a MFENCE, eg see CNC at

https://github.com/firedancer-io/firedancer/blob/70b0fabcba4efa67c6e3aa4b42ad9368825f66cc/src/tango/cnc/fd_cnc.c#L47

Reply:
Good catch.
Although FD_VOLATILE is incorrect there - volatile writes only guarantee in-order writes with respect to other volatiles. I'll use a fence

Reply:
The standard pattern in CNC is to use two MFENCEs, one on each side, along with the FD_VOLATILE